fieldsolver/derivatives.cpp

```diff
@@ -372,7 +372,10 @@ void calculateDerivativesSimple(
    for (int k=0; k<gridDims[2]; k++) {
       for (int j=0; j<gridDims[1]; j++) {
          for (int i=0; i<gridDims[0]; i++) {
-            if (technicalGrid.get(i,j,k)->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE) continue;
```

Comment:
So question: Is there any harm in calculating the derivatives for these cells as well? Inside the inner boundary they should be zero, but near the outer boundaries they might be nonzero. And then if we want to evaluate volumetric fields correctly there, we might need them. Removing a per-cell check could allow more vectorization, with very little actual cost?

Ditto for lines 519 and 665.

Reply:
We won't be using them though, right?

Reply:
Well, in constructing a volumetric value, there's a function `reconstructionCoefficients` which actually can use the derivative from the neighbouring cell. So... I'm not sure?

Reply:
This brings us to the uncomfortable question: Is there *any* mechanism in the code that could propagate information from these cells into the domain proper? Because I had assumed that no (so they are really just dead ballast, and can be ignored completely).

Reply:
No, reconstructionCoefficients is purely local?

Reply:
So I think in 2D in L2 boundary cells (outer edge of the wall) we probably set derivatives to zero so we don't access out of domain.
In 3D now we thus need to compute derivatives in the DNC/PADDING regions just in case, but only out to the penultimate layer? That would be sufficient once at startup, right? Until we go time-varying at inflow walls, there I suppose ideally we need the proper inflowing spatial profile/derivatives?? Ouch.

Reply:
Interestingly that `reconstructionCoefficients` https://github.com/fmihpc/vlasiator/blob/ae70882b72cd22ea9b8bcccca170757c1efaade0/fieldsolver/fs_common.cpp#L95 has a built-in way of managing right-at-the-edge cells, so it can also calculate derivatives there, it seems.

Reply:
So I do indeed agree that we should be able to compute derivatives throughout. And indeed we introduce the effect that the hinterland of the (inflow) boundary thus affects the volume averaged fields at the boundary (*not* inside-cells at least as long as the refinement is uniform along the wall, and the boundary is planar). Though this also means affecting at copy-boundaries, where the hinterland will also have initial values we don't want to keep affecting. That is in essence the same issue as with the time-varying inflow.
Therefore: I would advocate for zero-derivatives in DO_NOT_COMPUTE and OUTER_BOUNDARY_PADDING so we don't instead have "nastier" derivatives affecting us.

Reply:
Except that indeed as Markus was pointing out to me, we need to upkeep derivatives because of the "compensation term" in type 4 setups (vector potential modified dipole) storing a significant part in the perturbed field.

Reply:
However if we blindly remove this check, we'll run off the domain edges. We shouldn't call `calculateDerivatives()` for i,j,k == 0 or at the other end.

Reply:
Yep. Perhaps we could replicate the `reconstructionCoefficients` approach there?

Reply:
Did that in my branch, will be put online soon.